replication_state.go

```diff
@@ -0,0 +1,320 @@
+		return
+	}
+
+	// otherwise we are storing a round without finalization
+	// don't bother storing rounds that are older than the highest finalized round we know
+	if observed := r.finalizationRequestor.getHighestObserved(); observed != nil && observed.round >= round.GetRound() {
+		r.logger.Debug("Replication State received a finalized quorum round for a round we know is finalized.")
+		return
+	}
+
+	r.storeRound(round)
+	r.roundRequestor.receivedSignedQuorum(newSignedQuorum(round, r.myNodeID))
+	if round.EmptyNotarization != nil {
+		r.emptyRoundTimeouts.RemoveTask(round.GetRound())
+	}
+}
+
+// receivedFutureFinalization notifies the replication state a finalization was created in a future round.
+func (r *ReplicationState) ReceivedFutureFinalization(finalization *Finalization, nextSeqToCommit uint64) {
+	if !r.enabled {
+		return
+	}
+
+	signedSequence := newSingedQuorumFromFinalization(finalization, r.myNodeID)
+
+	// maybe this finalization was for a round that we initially thought only had notarizations
+	// remove from the round replicator since we now have a finalization for this round
+	r.deleteOldRounds(finalization.Finalization.BlockHeader.Round)
+
+	// potentially send out requests for blocks/finalizations in between
+	r.finalizationRequestor.maybeSendMoreReplicationRequests(signedSequence, nextSeqToCommit)
+}
+
+// receivedFutureRound notifies the replication state of a future round.
+func (r *ReplicationState) ReceivedFutureRound(round, seq, currentRound uint64, signers []NodeID) {
+	if !r.enabled {
+		return
+	}
+
+	if observed := r.finalizationRequestor.getHighestObserved(); observed != nil && observed.round >= round {
+		r.logger.Debug("Ignoring round replication for a future round since we have a finalization for a higher round", zap.Uint64("round", round))
+		return
+	}
+
+	sq := newSignedQuorumFromRound(round, seq, signers, r.myNodeID)
+	r.roundRequestor.maybeSendMoreReplicationRequests(sq, currentRound)
+}
+
+// ResendFinalizationRequest notifies the replication state that `seq` should be re-requested.
+func (r *ReplicationState) ResendFinalizationRequest(seq uint64, signers []NodeID) error {
+	if !r.enabled {
+		return nil
+	}
+
+	numSigners := int64(len(signers))
+	index, err := rand.Int(rand.Reader, big.NewInt(numSigners))
+	if err != nil {
+		return err
+	}
+
+	// because we are resending because the block failed to verify, we should remove the stored quorum round
+	// so that we can try to get a new block & finalization
+	delete(r.seqs, seq)
+	r.finalizationRequestor.sendRequestToNode(seq, seq, signers[index.Int64()])
+	return nil
+}
+
+// CreateDependencyTasks creates tasks to refetch the given parent digest and empty rounds. If there are no
+// dependencies, no tasks are created.
+func (r *ReplicationState) CreateDependencyTasks(parent *Digest, parentSeq uint64, emptyRounds []uint64) {
+	if parent != nil {
+		r.digestTimeouts.AddTask(*parent)
+		r.seqsToDigests[parentSeq] = *parent
```

Comment:
where are you removing from the map? This will eventually make the memory blow up, no?